test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,221 @@
+    const uint32_t size = clampSize(action.add_string());
+    auto string = std::make_unique<std::string>(size, 'a');
+    ctxt.strings_.emplace_back(std::move(string));
+    const uint32_t previous_length = target_buffer.length();
+    target_buffer.add(absl::string_view(*ctxt.strings_.back()));
+    ASSERT_EQ(previous_length, target_buffer.search(ctxt.strings_.back()->data(), size, 0));
+    break;
+  }
+  case test::common::buffer::Action::kAddBuffer: {
+    const uint32_t source_index = action.add_buffer() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    const std::string source_contents = source_buffer.toString();
+    const uint32_t previous_length = target_buffer.length();
+    target_buffer.add(source_buffer);
+    ASSERT_EQ(previous_length,
+              target_buffer.search(source_contents.data(), source_contents.size(), 0));
+    break;
+  }
+  case test::common::buffer::Action::kPrependString: {
+    const uint32_t size = clampSize(action.prepend_string());
+    auto string = std::make_unique<std::string>(size, 'a');
+    ctxt.strings_.emplace_back(std::move(string));
+    target_buffer.prepend(absl::string_view(*ctxt.strings_.back()));
+    ASSERT_EQ(0, target_buffer.search(ctxt.strings_.back()->data(), size, 0));
+    break;
+  }
+  case test::common::buffer::Action::kPrependBuffer: {
+    const uint32_t source_index = action.prepend_buffer() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    const std::string source_contents = source_buffer.toString();
+    target_buffer.prepend(source_buffer);
+    ASSERT_EQ(0, target_buffer.search(source_contents.data(), source_contents.size(), 0));
+    break;
+  }
+  case test::common::buffer::Action::kReserveCommit: {
+    const uint32_t previous_length = target_buffer.length();
+    const uint32_t reserve_length = clampSize(action.reserve_commit().reserve_length());
+    constexpr uint32_t reserve_slices = 16;
+    Buffer::RawSlice slices[reserve_slices];
+    uint32_t allocated_slices = target_buffer.reserve(reserve_length, slices, reserve_slices);
+    uint32_t allocated_length = 0;
+    for (uint32_t i = 0; i < allocated_slices; ++i) {
+      ::memset(slices[i].mem_, 'c', slices[i].len_);
+      allocated_length += slices[i].len_;
+    }
+    const uint32_t target_length =
+        std::min(allocated_length, action.reserve_commit().commit_length());
+    uint32_t shrink_length = allocated_length;
+    while (shrink_length > target_length) {
+      ASSERT(allocated_slices > 0);
+      const uint32_t available = slices[allocated_slices - 1].len_;
+      const uint32_t remainder = shrink_length - target_length;
+      if (available > remainder) {
+        slices[allocated_slices - 1].len_ -= remainder;
+        break;
+      }
+      shrink_length -= available;
+      --allocated_slices;
```

Comment:
If there are unfilled slices at the end, this loop probably should zero out their `len_`, rather than excluding them from the commit. (This wasn't explicitly documented in the `Buffer` interface in the past, but if the caller chops off any zero-length slices at the end of the array before calling `commit()`, it's possible that the underlying buffer implementation (whether evbuffer, the new design in PR 5441, or some future redesign) won't be able to reuse the corresponding internal slices for a future `reserve()` call. That will result in wasted memory.)

Reply:
So, https://github.com/envoyproxy/envoy/blob/master/include/envoy/buffer/buffer.h#L103 is then now stale? I.e. you should not vary the number of slices?

Reply:
As far as I can tell, that comment ("The number of slices can be different from the number obtained from reserve()") was only ever valid because of an implementation detail of libevent. The underlying libevent API function's documentation specifies that one should change the slice lengths to commit less data than was reserved; it doesn't guarantee that reducing the number of slices will work: https://github.com/libevent/libevent/blob/master/include/event2/buffer.h#L312